File: Java/Data Structures & Algorithms Module - Java/Revision/BinarySearchTree4/BST.java
/*
BST.java
Billy Cussen
01/02/2021
*/

public class BST<T extends Comparable<T>> {
    
    Node<T> root;

    public BST(){
        this.root = null;
    }

    public boolean isEmpty(){
        return this.root == null;
    }

    public void insert(T element){
        if(isEmpty()){
            this.root = new Node<T>(element);
        } else{
            insertHelper(element, this.root);
        }
    }

    private void insertHelper(T element, Node<T> current){
        if(element.compareTo(current.element)==-1){
            if(current.left == null){
                current.left = new Node<T>(element);
            } else{
                insertHelper(element, current.left);
            }
        } else{
            if(current.right == null){
                current.right = new Node<T>(element);
            } else{
                insertHelper(element, current.right);
            }
        }
    }

    public boolean contains(T element){
        if(isEmpty()){
            return false;
        }
        return containsHelper(element, this.root);
    }

    private boolean containsHelper(T element, Node<T> current){
        if(current==null){
            return false;
        }

        if(element.compareTo(current.element)==0){
            return true;
        } else if(element.compareTo(current.element) == -1){
            return containsHelper(element, current.left);
        } else{
            return containsHelper(element, current.right);
        }
    }

    public int size(){
        if(isEmpty()){
            return 0;
        }
        return sizeHelper(this.root);
    }

    private int sizeHelper(Node<T> current){
        if(current == null){
            return 0;
        }
        return 1 + sizeHelper(current.left)+sizeHelper(current.right);
    }

    public T max(){
        if(isEmpty()){
            return null;
        }
        return maxHelper(this.root);
    }

    private T maxHelper(Node<T> current){
        if(current.right == null){
            return current.element;
        } else{
            return maxHelper(current.right);
        }
    }

    public T min(){
        if(isEmpty()){
            return null;
        }
        return minHelper(this.root);
    }

    private T minHelper(Node<T> current){
        if(current.left == null){
            return current.element;
        } else{
            return minHelper(current.left);
        }
    }

    public Node<T> findNode(T element){
        if(isEmpty()){
            return null;
        } else{
            return findNodeHelper(element, this.root);
        }
    }

    private Node<T> findNodeHelper(T element, Node<T> current){
        if(current == null){
            return null;
        }

        if(element.compareTo(current.element)==0){
            return current;
        } else if(element.compareTo(current.element)==-1){
            return findNodeHelper(element, current.left);
        } else{
            return findNodeHelper(element, current.right);
        }
    }

    public Node<T> findParent(T element){
        if(isEmpty()||this.root.element==element){
            return null;
        }
        return findParentHelper(element, this.root);
    }

    private Node<T> findParentHelper(T element, Node<T> current){
        if(current == null){
            return null;
        }

        if(element.compareTo(current.element)==-1){
            if(element.compareTo(current.left.element)==0){
                return current;
            } else{
                return findParentHelper(element, current.left);
            }
        } else{
            if(element.compareTo(current.right.element)==0){
                return current;
            } else{
                return findParentHelper(element, current.right);
            }
        }
    }

    public int depth(){
        if(isEmpty()){
            return 0;
        }
        return depthHelper(this.root);
    }

    private int depthHelper(Node<T> current){
        if(current == null){
            return 0;
        }

        int left = 1 + depthHelper(current.left);
        int right = 1 + depthHelper(current.right);
        if(left >= right){
            return left;
        } else{
            return right;
        }
    }

    public void inOrder(){
        if(isEmpty()){
            System.out.println("Can't traverse, its empty!");
        } else{
            System.out.println("***BST INORDER TRAVERSAL***");
            inOrderHelper(this.root);
            System.out.println("***END***");
        }
    }

    private void inOrderHelper(Node<T> current){
        if(current==null){
            return;
        }

        inOrderHelper(current.left);
        System.out.println(current.toString());
        inOrderHelper(current.right);
    }

    public boolean remove(T element){

        Node<T> toRemove = findNode(element);
        Node<T> parentNode = findParent(element);

        if(isEmpty() || contains(element)==false || this.root.element.compareTo(element)==0){
            return false;
        } else if(toRemove.left == null && toRemove.right == null){
            if(toRemove.element.compareTo(parentNode.element)==-1){
                toRemove.left = null;
            } else{
                toRemove.right = null;
            } 
            return true;
        } else if(toRemove.left != null && toRemove.right == null){
            if(toRemove.element.compareTo(parentNode.element)==-1){
                parentNode.left = toRemove.left;
            } else{
                parentNode.right = toRemove.left;
            }
            return true;
        } else if(toRemove.left == null && toRemove.right != null){
            if(toRemove.element.compareTo(parentNode.element)==-1){
                parentNode.left = toRemove.right;
            } else{
                parentNode.right = toRemove.right;
            }
            return true;
        } else if(toRemove.left != null && toRemove.right != null){
            T max = maxHelper(toRemove.left);
            Node<T> replacement = findNode(max);
            Node<T> replacementParent = findParent(max);
            replacementParent.right = replacement.left;
            toRemove.element = replacement.element;
            return true;
        } 
        return true;
    }

}

File: Python/Data Structures & Algorithms Module - Python/Revision/BST3/BST.py
"""
BST.py
Billy Cussen
03/02/2021
"""
from Node import Node

class BST:

    def __init__(self):
        self.root = None

    def isEmpty(self):
        return self.root == None

    def insert(self, element):
        if self.isEmpty():
            self.root = Node(element)
        else:
            self.insertHelper(element, self.root)

    def insertHelper(self, element, current):
        if (element > current.element)-(element < current.element)==-1:
            if current.left == None:
                current.left = Node(element)
            else:
                self.insertHelper(element, current.left)
        else:
            if current.right == None:
                current.right = Node(element)
            else:
                self.insertHelper(element, current.right)

    def inOrder(self):
        if self.isEmpty():
            return "Its Empty!"
        result = "***BST IN ORDER TRAVERSAL***\n"
        result+= self.inOrderHelper(self.root)
        result+= "***END***"
        return result

    def inOrderHelper(self, current):
        result=""
        if current == None:
            return ""
        result+= self.inOrderHelper(current.left)
        result+= current.toString()+"\n"
        result+= self.inOrderHelper(current.right)
        return result

    def contains(self, element):
        if self.isEmpty():
            return False
        return self.containsHelper(element, self.root)

    def containsHelper(self, element, current):
        if current == None:
            return False
        elif (element > current.element)-(element < current.element)==0:
            return True
        elif (element > current.element)-(element < current.element)==-1:
            return self.containsHelper(element, current.left)
        else:
            return self.containsHelper(element, current.right)

    def findNode(self, element):
        if self.isEmpty() or self.contains(element)==False:
            return None
        return self.findNodeHelper(element, self.root)

    def findNodeHelper(self, element, current):
        if (element > current.element)-(element < current.element)==0:
            return current
        elif (element > current.element)-(element < current.element)==-1:
            return self.findNodeHelper(element, current.left)
        else:
            return self.findNodeHelper(element, current.right)

    def findParent(self, element):
        if self.isEmpty() or self.contains(element)==False or self.root.element == element:
            return None
        return self.findParentHelper(element, self.root)

    def findParentHelper(self, element, current):
        if(element > current.element)-(element < current.element)==-1:
            if (element > current.left.element)-(element < current.left.element)==0:
                return current
            else:
                return self.findParentHelper(element, current.left)
        else:
            if (element > current.right.element)-(element < current.right.element)==0:
                return current
            else:
                return self.findParentHelper(element, current.right)

    def remove(self, element):
        if self.isEmpty() or self.contains(element)==False:
            return False
        
        toRemove = self.findNode(element)
        parentNode = self.findParent(element)

        if toRemove.left == None and toRemove.right == None:
            if (toRemove.element > parentNode.element)-(toRemove.element < parentNode.element) == -1:
                parentNode.left = None
            else:
                parentNode.right = None
            return True
        elif toRemove.left != None and toRemove.right == None:
            if (toRemove.element > parentNode.element)-(toRemove.element < parentNode.element) == -1:
                parentNode.left = toRemove.left
            else:
                parentNode.right = toRemove.left
            return True
        elif toRemove.left == None and toRemove.right != None:
            if (toRemove.element > parentNode.element)-(toRemove.element < parentNode.element) == -1:
                parentNode.left = toRemove.right
            else:
                parentNode.right = toRemove.right
            return True
        return False

    
